File: COA software/frontend/src/types/whatIf.ts
export interface ScenarioParameters {
  base_plan_id?: string
  base_plan_version?: number
  corridor_id?: string
  block_start: string
  block_duration_minutes: number
  block_end?: string
  passenger_traffic_multiplier: number
  goods_forecast_rate: number
  selected_tasks: string[]
  task_duration_overrides: Record<string, number>
  task_priority_overrides: Record<string, number>
  available_window_start?: string
  available_window_end?: string
}

export interface KpiDelta {
  baseline: number
  scenario: number
  diff: number
  pct_change: number
  status: 'IMPROVED' | 'UNCHANGED' | 'WORSE' | 'CRITICAL'
}

export interface ScenarioExplanation {
  what_changed: string
  what_happened: string
  why: string[]
  recommendation: string
}

export interface AlternativeRecommendation {
  window: string
  corridor_id: string
  expected_train_delay: number
  conflicts: number
  savings_vs_scenario: {
    delay_reduced_minutes: number
    score_improvement: number
  }
  rationale: string
}

export interface ScenarioResult {
  scenario_id: string
  status: string
  executed_at: string
  baseline_metrics: Record<string, any>
  scenario_metrics: Record<string, any>
  deltas: Record<string, KpiDelta>
  explanation: ScenarioExplanation
  alternative_recommendation?: AlternativeRecommendation | null
  score: number
}

export interface SimulationScenarioItem {
  id: string
  name: string
  description?: string
  scenario_type: string
  configuration?: {
    base_plan_id?: string
    base_plan_version?: number
    parameters: ScenarioParameters
    baseline_metrics?: Record<string, any>
    status: 'DRAFT' | 'READY' | 'RUNNING' | 'COMPLETED' | 'FAILED'
    results?: ScenarioResult | null
  }
  created_by?: string
  created_at?: string
  updated_at?: string
}

File: COA software/backend/app/schemas/what_if.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime


class ScenarioCreateRequest(BaseModel):
    name: str = Field(..., min_length=2, max_length=128, description="Scenario name")
    description: Optional[str] = Field(None, max_length=500, description="Scenario description")
    base_plan_id: Optional[str] = Field(None, description="Base Plan ID snapshot")
    corridor_id: Optional[str] = Field(None, description="Target corridor ID")
    scenario_type: Optional[str] = Field(default="WHAT_IF_EXPERIMENT")
    parameters: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Scenario parameter overrides")


class ScenarioUpdateRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    parameters: Optional[Dict[str, Any]] = None


class ScenarioDetailResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    scenario_type: str
    configuration: Optional[Dict[str, Any]] = None
    created_by: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ScenarioValidationResponse(BaseModel):
    scenario_id: str
    is_valid: bool
    conflicts: List[str] = []
    validation_timestamp: str


class ScenarioRunResultResponse(BaseModel):
    scenario_id: str
    status: str
    executed_at: str
    baseline_metrics: Dict[str, Any]
    scenario_metrics: Dict[str, Any]
    deltas: Dict[str, Any]
    explanation: Dict[str, Any]
    alternative_recommendation: Optional[Dict[str, Any]] = None
    score: float


class ScenarioCompareRequest(BaseModel):
    scenario_ids: List[str]


class ScenarioCompareResponse(BaseModel):
    comparison: List[Dict[str, Any]]
    best_option: Optional[Dict[str, Any]] = None
